osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/messaging/processors/KeysRequestMessageProcessor.java

```diff
@@ -0,0 +1,108 @@
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.adapter.protocol.dlms.infra.messaging.processors;
+
+import java.io.Serializable;
+import java.time.Duration;
+import javax.jms.JMSException;
+import javax.jms.ObjectMessage;
+import lombok.extern.slf4j.Slf4j;
+import org.opensmartgridplatform.adapter.protocol.dlms.application.services.ConfigurationService;
+import org.opensmartgridplatform.adapter.protocol.dlms.application.services.DeviceKeyProcessingService;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.DlmsDevice;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.factories.DlmsConnectionManager;
+import org.opensmartgridplatform.adapter.protocol.dlms.exceptions.DeviceKeyProcessAlreadyRunningException;
+import org.opensmartgridplatform.adapter.protocol.dlms.infra.messaging.DeviceRequestMessageProcessor;
+import org.opensmartgridplatform.adapter.protocol.dlms.infra.messaging.DeviceRequestMessageSender;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.SetKeysRequestDto;
+import org.opensmartgridplatform.shared.exceptionhandling.OsgpException;
+import org.opensmartgridplatform.shared.infra.jms.MessageMetadata;
+import org.opensmartgridplatform.shared.infra.jms.MessageType;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@Slf4j
+public class KeysRequestMessageProcessor extends DeviceRequestMessageProcessor {
+
+  @Autowired private ConfigurationService configurationService;
+
+  @Autowired private DeviceRequestMessageSender deviceRequestMessageSender;
+
+  @Autowired private DeviceKeyProcessingService deviceKeyProcessingService;
+
+  public KeysRequestMessageProcessor(final MessageType messageType) {
+    super(messageType);
+  }
+
+  @Override
+  protected Serializable handleMessage(
+      final DlmsConnectionManager conn,
+      final DlmsDevice device,
+      final Serializable requestObject,
+      final MessageMetadata messageMetadata)
+      throws OsgpException {
+
+    this.assertRequestObjectType(SetKeysRequestDto.class, requestObject);
+
+    final SetKeysRequestDto keySet = (SetKeysRequestDto) requestObject;
+    this.configurationService.replaceKeys(conn, device, keySet, messageMetadata);
+
+    return null;
+  }
+
+  @Override
+  public void processMessage(final ObjectMessage message) throws JMSException {
+    final MessageMetadata messageMetadata = MessageMetadata.fromMessage(message);
+    final Serializable messageObject = message.getObject();
+
+    log.debug(
+        "Checking if not already processing key for device {} before processing {} request message",
+        messageMetadata.getDeviceIdentification(),
+        messageMetadata.getMessageType());
+
+    try {
+      final DlmsDevice device = this.domainHelperService.findDlmsDevice(messageMetadata);
```

Comment:
The method `findDlmsDevice(MessageMetadata)` is a method with potential side effects - and yes it is badly named, because it hides those actions behind an unrevealing name - making calls to a session provider to get an IP address. This is already done where the IP address is needed, and should not be repeated unless there is a need to. There are other helper methods that do not have these side effects. For instance in the `domainHelperService` there is another `findDlmsDevice(String deviceIdentification)` that just does the lookup.

Reply:
if the `deviceIdentification` is already in the `messageMetadata` and `deviceIdentification` is all that is needed then the call to `findDlmsDevice` is not needed at all.